src/app/mod.rs

```diff
@@ -356,6 +482,78 @@ impl App {
         Ok(proof.into())
     }
 
+    /// # Errors
+    ///
+    /// Will return `Err` if the provided index is out of bounds.
+    #[instrument(level = "debug", skip(self))]
+    pub async fn inclusion_proof_v2(
+        &self,
+        commitment: Hash,
+    ) -> Result<(Field, Field, semaphore_rs::poseidon_tree::Proof), InclusionProofV2Error> {
+        if self.identity_validator.is_initial_leaf(&commitment) {
+            return Err(InclusionProofV2Error::InvalidCommitment);
+        }
+
+        if !self.identity_validator.is_reduced(commitment) {
+            warn!(
+                ?commitment,
+                "The provided commitment is not an element of the field."
+            );
+            return Err(InclusionProofV2Error::UnreducedCommitment);
+        }
+
+        let mut tx = self
+            .database
+            .begin_tx(IsolationLevel::RepeatableRead)
+            .await?;
+
+        let item = match tx.get_tree_item(&commitment).await? {
+            None => {
+                let unprocessed = tx.get_unprocessed_commitment(&commitment).await?;
+                if unprocessed.is_some() {
+                    return Err(InclusionProofV2Error::UnprocessedCommitment);
+                } else {
+                    return Err(InclusionProofV2Error::CommitmentNotFound);
+                }
+            }
+            Some(tree_item) => tree_item,
+        };
+
+        let latest_at_index = match tx.get_tree_item_by_leaf_index(item.leaf_index).await? {
+            None => {
+                // This should never happen as last query was returning row with that value
+                return Err(InclusionProofV2Error::CommitmentNotFound);
+            }
+            Some(tree_item) => tree_item,
+        };
+
+        if latest_at_index.element == Hash::ZERO {
+            return Err(InclusionProofV2Error::DeletedCommitment);
+        }
+
+        let tree_state = self.tree_state()?;
+        if tree_state.latest_tree().get_last_sequence_id() < item.sequence_id {
+            // If tree change was not yet applied to latest tree we treat it as it would be
+            // unprocessed.
+            return Err(InclusionProofV2Error::UnprocessedCommitment);
+        }
+
+        // If tree change was applied to latest tree we look for inclusion proof
+        // todo(piotrh): is it ok to use latest tree here?
+        let (leaf, root, proof) = self
+            .tree_state()?
+            .get_latest_tree()
+            .get_leaf_and_proof(item.leaf_index);
+
+        if leaf != commitment {
+            return Err(InclusionProofV2Error::InvalidCommitment);
```

Comment:
This should be an internal server error of some sort. When would this ever happen?